libs/wingc/src/closure_transform.rs

```diff
@@ -0,0 +1,306 @@
+	inflight_counter: usize,
+	// Stores the list of class definitions that need to be added to the nearest scope
+	class_statements: Vec<Stmt>,
+	// Track the statement index of the nearest statement we're inside so that
+	// newly-inserted classes will have access to the correct state
+	nearest_stmt_idx: usize,
+}
+
+impl ClosureTransformer {
+	pub fn new() -> Self {
+		Self {
+			curr_phase: Phase::Preflight,
+			inside_scope_with_this: false,
+			inflight_counter: 0,
+			class_statements: vec![],
+			nearest_stmt_idx: 0,
+		}
+	}
+}
+
+impl Fold for ClosureTransformer {
+	fn fold_scope(&mut self, node: Scope) -> Scope {
+		let mut statements = vec![];
+
+		// If we are inside a scope with "this", add define `let __parent_this = this` which can be
+		// used by the newly-created preflight classes
+		if self.inside_scope_with_this {
+			let parent_this_name = Symbol::global(PARENT_THIS_NAME); // TODO: can we use a span?
+			let this_name = Symbol::global("this"); // TODO: can we use a span?
+			let parent_this_def = Stmt {
+				kind: StmtKind::VariableDef {
+					reassignable: false,
+					var_name: parent_this_name,
+					initial_value: Expr {
+						kind: ExprKind::Reference(Reference::Identifier(this_name)),
+						span: WingSpan::default(),          // TODO: can we use a span?
+						evaluated_type: RefCell::new(None), // thank god we have type reference
+					},
+					type_: None,
+				},
+				span: WingSpan::default(), // TODO: can we use a span?
+				idx: 0,
+			};
+
+			statements.push(parent_this_def);
+		}
+
+		for stmt in node.statements {
+			let old_nearest_stmt_idx = self.nearest_stmt_idx;
+			self.nearest_stmt_idx = stmt.idx;
+			let new_stmt = self.fold_stmt(stmt);
+			self.nearest_stmt_idx = old_nearest_stmt_idx;
+
+			// Add any new statements we have accumulated that define temporary classes
+			statements.append(&mut self.class_statements);
+
+			// Push the folded statement last. This way it has access to all of the temporary
+			// classes we have defined.
+			statements.push(new_stmt);
+
+			// Reset scope-specific state
+			self.class_statements.clear();
+		}
+
+		Scope {
+			statements,
+			span: node.span,
+			env: node.env,
+		}
+	}
+
+	fn fold_function_definition(&mut self, node: FunctionDefinition) -> FunctionDefinition {
+		let prev_phase = self.curr_phase;
+		self.curr_phase = node.signature.phase;
+		let new_node = fold::fold_function_definition(self, node);
+		self.curr_phase = prev_phase;
+		new_node
+	}
+
+	fn fold_class(&mut self, node: Class) -> Class {
+		Class {
+			name: self.fold_symbol(node.name),
+			fields: node
+				.fields
+				.into_iter()
+				.map(|field| self.fold_class_field(field))
+				.collect(),
+			methods: node
+				.methods
+				.into_iter()
+				.map(|(name, def)| {
+					let new_symbol = self.fold_symbol(name);
+
+					let new_def = if !def.is_static {
+						let prev_inside_scope_with_this = self.inside_scope_with_this;
+						self.inside_scope_with_this = true;
```

Comment:
I have a feeling we can move this logic into `fold_function_definition` which might be cleaner.
Same for `fold_constructor`.
If we do this, we might be able to keep the default implementation of `fold_class`.

Reply:
I think the idea sounds reasonable... unfortunately when I tried this now, I encountered an issue because when `is_static: false` on a FunctionDefinition does not imply it is an instance method (for example, an anonymous preflight function like `() => {}` is not currently treated as static in the compiler). I wasn't sure if we should change that, so I'll leave this logic for now

Reply:
I see. It actually seems like a bug, if we change `build_function_definition` in _parser.rs_ to correctly set `is_static` to `true` when called from `build_anonymous_closure` it might resolve this. You may try or leave it for a future improvement...

Reply:
OK - updated